plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,537 @@
+	if (back_data_addr == AG_UPDATE_STATUS)
+		back_data_value = back_data->data[0];
+	else
+		back_data_value = fu_memread_uint16(back_data->data + 4, G_BIG_ENDIAN);
+	return back_data_value;
+}
+
+static gboolean
+algoltek_device_en(FuAlgoltekUsbDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 3;
+	buf->data[1] = FU_ALGOLTEK_CMD_EN;
+	buf->data[10] = ~fu_sum8(buf->data, buf->len) + 1;
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_EN,
+					   0,
+					   0,
+					   buf,
+					   buf->data[0],
+					   error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_rst(FuAlgoltekUsbDevice *self, guint8 number, GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 4;
+	buf->data[1] = FU_ALGOLTEK_CMD_RST;
+	buf->data[2] = number;
+	buf->data[10] = ~fu_sum8(buf->data, buf->len) + 1;
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_RST,
+					   0,
+					   0,
+					   buf,
+					   buf->data[0],
+					   error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_wrr(FuAlgoltekUsbDevice *self, int address, int input_value, GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 7;
+	buf->data[1] = FU_ALGOLTEK_CMD_WRR;
+	fu_memwrite_uint16(buf->data + 2, address, G_BIG_ENDIAN);
+	fu_memwrite_uint16(buf->data + 4, input_value, G_BIG_ENDIAN);
+	buf->data[10] = ~fu_sum8(buf->data, buf->len) + 1;
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_WRR,
+					   0,
+					   0,
+					   buf,
+					   buf->data[0],
+					   error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_ISP(FuAlgoltekUsbDevice *self,
+		    GBytes *blob_isp,
+		    guint address,
+		    FuProgress *progress,
+		    GError **error)
+{
+	g_autoptr(GPtrArray) chunks_isp = NULL;
+	const guint8 *isp_data = NULL;
+	gsize isp_data_size = 0;
+	guint8 basic_data_size = 5;
+	guint data_addr = address;
+
+	isp_data = g_bytes_get_data(blob_isp, &isp_data_size);
+	chunks_isp = fu_chunk_array_new(isp_data, isp_data_size, 0x0, 0x0, 64 - basic_data_size);
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks_isp->len);
+
+	for (guint i = 0; i < chunks_isp->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks_isp, i);
+		g_autoptr(GByteArray) part_isp_data = g_byte_array_new();
+
+		fu_byte_array_append_uint8(part_isp_data,
+					   basic_data_size + fu_chunk_get_data_sz(chk));
+		fu_byte_array_append_uint8(part_isp_data, FU_ALGOLTEK_CMD_ISP);
+		fu_byte_array_append_uint16(part_isp_data, data_addr, G_BIG_ENDIAN);
```

Comment:
can you use `fu_chunk_get_address(chk)` by using the start address in `fu_chunk_array_new()`?

Reply:
Updated! Use `fu_chunk_get_address(chk)` and assign the start address in `fu_chunk_array_new()`.